### PR TITLE
Update mkimage to add support gpt as select build option

### PR DESCRIPTION
**Pull request by wqmeng**
# use user forced disk label if available. Will be useful for big USB HDD over 2TB formatting with gpt.

I tested on my build on a Amlogic S905X4 TV Box with a USB 5TB SATA Seagate HDD, and work with gpt disk formatting. 

Build with 
`PROJECT=Amlogic-ce DEVICE=Amlogic-ng ARCH=aarch64 DISTRO=EmuELEC DISK_LABEL=gpt make image
`

When done, flash the img to the 5TB USB HDD. And plug it to the TV Box USB port. When boot up, I suppose the SATA HDD is slow for init to boot up,  And force to add a TF card (not boot-able) to the TF port, then the BOX boot firstly from the TF card, then check the USB port after then, the 5TB HDD is ready and boot up successfully. Without the TF card in the TV BOX, it can not boot up. 

When done, the EmuELEC will create a 2GB EmuELEC, 6GB STORAGE, 2TB FAT32 EEROMS partitions. I reformat the 2TB FAT32 partition to a 4.5TB exFAT EEROMS partition and it works well.

The only shortcoming is having to keep the TF card in the TV BOX. But I think if you have a speedy USB 5TB SSD, it will boot up correctly without the TF card.

Thank you.

### CODE REVIEW

**Comment by 7Ji** [comment]
Several things I need to make clear here:
 - You are adding a variable to set yet there's nothing to check if it is correct, previously ``DISK_LABEL`` is set depending on ``bootloader``, and it could only be either ``msdos`` or ``gpt``. With this change, you can set it to something else and the build process will go on normally yet fail badly at the end.
 - Boot priority is always ``USB->TF->eMMC->store boot``, where ``store boot`` is usually the Android sitting on eMMC but it's not guaranteed. It's impossible that TF comes earlier than USB (Unless you've adjusted u-boot envs by yourself). I'm doubting It's actually TF that's loaded by u-boot (yet it then used partitions on USB drive, see below for how partitions are chosen for boot)
 - GPT support on Amlogic u-boot is optional and not guaranteed, because it's barely needed for the drive that'll be used
 - There should be only one partition labeled as ``EMUELEC``, one partition labeled as ``STORAGE``, and one partition labeled as ``EEROMS``, because the init process mounts these partitions by label. You should **never** mix multiple boot drives, the partitions being mounted will be messed up, and it's hard to guess which combination is used (e.g. You might load ``cfgload`` from TF, yet mounts ``EMUELEC`` from one of the USB drive, and ``STORAGE`` from another USB drive, if you have that many different drives). Effectively, you should **only connect one drive that is bootable and contains the above partitions to the box** (I don't think many will mix multiple EE parts, but mixing CE and EE can also bring in problems due to parts being mounted by label, as CE also has a ``STORAGE`` part)
 - FAT32, per the spec, has a max fs size of 2TiB (This can be larger if a non-default sector size is used).
 - Filesystem != Partition. In your case it was a 4.5 TiB part 3 that's created, and a FAT32 fs is created on it. The FS does not take up the entire partition due to its size limitation. You re-formatting changed that 2TiB FAT32 FS on the 4.5TiB partition to a 4.5TiB exFAT FS on the 4.5TiB partition
 - If you want to use a large spin drive, you should use it as external ROMs drive. It's always suggested to use a microSD card as the boot drive (or a UAS SSD, or some other fast USB3.0 drive). You can create a better solution than yours with the current dev branch and the upcoming 4.6 release (it's also possible in current versions, but you can't mix multiple ROMs drives as below):
   - Flash the image as-is on your microSD card, do the DTB replacing work as always in the ``EMUELEC`` partition
   - (Optional) Create a file named ``ee_fstype``(no extension) inside the ``EMUELEC`` partition, write ``no`` in it, this will disable the creation of ``EEROMS`` partition: the ``STORAGE`` partition will take all of the remaining space. (``no`` is a special value added lately, you can also set the FS you want to use for ``EEROMS`` in it: ``ext4``, ``exfat``, ``ntfs``, or ``fat32``.  Only ``no`` is new, the remaining works in older versions)
   - Format your external ROMs with any FS you like (NTFS, exFAT, ext4, xfs, FAT, etc) as long as its supported by EE's kernel, and **do not** label the partition as ``EMUELEC``, ``STORAGE``, or ``EEROMS``
   - Structure your external ROMs partition in the following layout:
      ```
      (root)
           L  roms/
                 L nes/
                 L snes/
                 L neogeo/
                ...
                 L emuelecroms
      ```
    - If you want to use all systems in this partition (the whole ``roms`` folder), keep the ``emuelecroms`` (no extension) file empty, otherwise, write the systems you want to use in it, one system per line, e.g.
      ```
      neogeo
      ps
      ```
      This will only use the ROMs for neogeo and ps from this drive. You can mix as many different ROMs drive as possible however you like it. The only thing you need to remember is that only one drive can provide ROMs from the whole ``roms`` folder, and only one drive can provide ROMs for each system.
 - Before creating your commit, sync with our ``dev`` branch, your PR branch has merge conflicts, which is not nice and impossible to merge.

**Comment by shantigilbert** [comment]
Sorry but this change won't be merged, this is a specific case, and it also requires an image to be created for this specific case which will not happen when we do releases. If a user wants this specific case (like yourself) you can add the changes and built from scratch.

**Comment by 7Ji** [comment]
@wqmeng 
And I need to add to the point why creating a GPT image is not needed even for yourself: You don't even need an image to create an CoreELEC/EmuELEC boot drive in the first place, the image just happens to come with all required files (in fact, more than required files, e.g. you don't need many of those DTBs) and patition layout on it for the ease of use but is not the only way a boot drive can be created. Since you already dig to this depth and want maximum customization, you can follow my article on how to create a boot drive with the ``.tar`` file that's usually used for update, with freely set partition sizes:
https://7ji.github.io/elec/2022/07/08/elec_tarboll_to_image.html

**Comment by wqmeng** [comment]
@7Ji 

Yes, it will be much safe to force the DISK_LABEL to be msdos or gpt, but who knows someday there will be more than just mbr and gpt for formatting? I just provide this for a select options when do custom build.

I have read from someone proved that the boot order is 
`TF-> USB -> eMMC -> ? `, That why so many video UPPers suggest their viewers to use a TF-CARD as their EmuELEC disk. 
And in my case my TF-card can be empty, it does not including any partitions from EE. Just used to for my TV BOX wait as long as my USB HDD is read to boot up.  And it was proved that works. 

Though you can use a USB SSD as a storage, but the cost of a 5TB SSD is still expensive than a cheap 5TB USB HDD.  And my 5TB HDD is very fast enough for EE usage.  Fast boot, big storage 4.5TB free space, fast user menu interface without any delay,  also maybe fast game loading. 

In other words, there is no any worse with your suggestion. 

I can also use a small USB disk or a TF-card for EE, and use the 5TB HDD only as a external storage with emuelecroms file on it. But my USB disk or the TF-card in-hand is slow when boot up EE, I do not want to buy a new faster small disk which only has 64GB or 128GB space for EE than I get a 5TB USB HDD instead with such much space. 

@shantigilbert 
What I do here just to verify the HDD or SSD over 2TB with gpt can work with EE very well, if you like this build. And my image can also work on my 8GB USB Disk with a gpt formatting, there is no much disadvantage comparing the mbr formatting. 

If the bootloader on an old box can not read gpt, this build will surly does not work. Otherwise, it is ready for USB over 2TB. 

Thank you

**Comment by shantigilbert** [comment]
> If the bootloader on an old box can not read gpt, this build will surly does not work. Otherwise, it is ready for USB over 2TB.

And here lies the problem, this has not been fully tested, hence being a SPECIFIC case. Unless you can provide a way to make this at runtime without affecting other devices, then this PR won't be merged. What will happen when users run into issues? what would the solutions be? when most s905 devices will most likely fail (they have very old boot loaders and we are not 100% sure they support GPT)? do we have to provide 2 images just for such a dedicated solution? 

I know this solves an issue for you, but it cannot be merged (and used) as is on the main image, the way it is right now is compatible with all the devices we support, your way is not.  But again you can still use this on your own image.

**Comment by wqmeng** [comment]
Hello,

Nobody request a gpt copy of img from EE official releases,  This is a build option, only provide a quick custom gpt build solution by adding DISK_LABLE=gpt in the building env.

Just a message, if you do not merge it, it will be okay.

Thank you.

**Comment by 7Ji** [comment]
> but who knows someday there will be more than just mbr and gpt for formatting? I just provide this for a select options when do custom build.

"Future-proof" should not be the reason to not sanity-check. E.g., we have ``Amlogic-ng`` and ``Amlogic-old`` devices under ``Amlogic-ce`` project, and you can't set the device to anything else when building project ``Amlogic-ce``, do you think this is not future-prrof and limits our future expansion to other ``Amlogic-ce`` devices? The answer is no. What you should do is not "limiting" the ``DISK_LABEL`` to ``msdos`` and ``gpt`` forever, but for now when there should only be ``msdos`` and ``gpt``. 

If you didn't read the sources enough, the tool we use to create the partition table on the image is GNU/parted, and it supports a lot of different labels: "aix", "amiga", "bsd", "dvh", "gpt", "loop", "mac", "ms‐dos", "pc98", or "sun" (source: ``man parted.8``, line 67). And you can set the ``DISK_LABEL`` to any of those other than ``ms-dos`` and ``gpt`` which won't break the build process yet will result to an image that does not work **at all**. And you should definitely do sanity-check to guarantee the ``DISK_LABEL`` is only either ``ms-dos``, ``msdos``(an alias used in GNU/parted for ``ms-dos``) or ``gpt``

> the boot order is
``TF-> USB -> eMMC -> ? ``

Sorry I mis-remembered the boot order, because all my primary CE/EE setups use eMMC as boot drive and I only alter to use USB when testing. When I have Android, CE, EE all three co-existing on eMMC, I really don't try any other boot options

> Just used to for my TV BOX wait as long as my USB HDD is read to boot up. And it was proved that works.

If you want the box to wait longer, but without a microSD card occupying the slot, and waste time on non-existing boot option, you can change the bootcmd here https://github.com/EmuELEC/EmuELEC/blob/dev/projects/Amlogic-ce/devices/Amlogic-old/bootloader/scripts/aml_autoscript.src#L8
from
```
setenv bootcmd 'if test ${bootfromnand} = 1; then setenv bootfromnand 0; saveenv; else run bootfromsd; run bootfromusb; fi; run storeboot'
```
to
```
setenv bootcmd 'if test ${bootfromnand} = 1; then setenv bootfromnand 0; saveenv; else usleep 1000000; run bootfromusb; fi; run storeboot'
```
where usleep is a command to sleep in microsecond (and 10e6 microseconds means 1 second)
You'll need to hold the reset button or use ``Reboot to LibreELEC`` tool or run ``reboot update`` from Android to put the box to update mode to re-read the ``aml_autoscript`` though.

**Comment by wqmeng** [comment]
Hello @7Ji 

Change the uboot script file should work after re-flash the eMMC OS or install the uboot bootloader to the eMMC storage, If only burn the img on a HDD or USB Disk, it will not work. I add the tf-card to force the bootloader of the original TV box to wait a bit longer for the HDD is ready to boot.  The trick is simple and do not need to flash the TV BOX eMMC storage. 

Yes, we could write the DISK_LABEL code more safe. Also something could improve such as use full of the DISK space to extend the EEROMS partition instead of limit it only 2TB,  Also use exFAT not FAT32 maybe.

Thank you.

**Comment by 7Ji** [comment]
> use full of the DISK space to extend the EEROMS partition instead of limit it only 2TB, Also use exFAT not FAT32 maybe.

It **is** using the full space, the fs-resize script will 100% resize the last partition to use all of the remaining space as that's how I wrote it:
https://github.com/EmuELEC/EmuELEC/blob/3c3012546004c8b1287a69b16c6ed29def60f38e/packages/sysutils/busybox/scripts/fs-resize#L52
It is limited to 2TiB because a FAT32 **filesystem** can only be 2TiB at max, so when a FAT32 **filesystem** is created on a 4.5TiB **partition**, it will only be 2TiB, but re-formatting the **filesystem** to other without the limitation (exFAT, NTFS, ext4, etc) will use again the remaining space. If the partition is 2TiB, you can't just re-format it to be a 4.5TiB exFAT, you'll have to resize it before the creation of the exFAT filesystem, which is not the case

> Also use exFAT not FAT32 maybe.

There is always the choice to use a user-defined FS when resizing the disk and creating the ``EEROMS`` partition:
https://github.com/EmuELEC/EmuELEC/blob/3c3012546004c8b1287a69b16c6ed29def60f38e/packages/sysutils/busybox/scripts/fs-resize#L78

This is documented way way before in the v4.2 release log:
https://github.com/EmuELEC/EmuELEC/releases/tag/v4.2

To issue the fs-resize script to create the ``EEROMS`` partition as ``exFAT``, just create a file ``ee_fstype`` in the ``EMUELEC`` partition (the partition where you adjust the DTB) after you burn the image to a drive, and write ``exfat`` in it.

>  do not need to flash the TV BOX eMMC storage.

I gave the advice to change the script before compiling because modifying u-boot env afterwards is dangerous, but you can do that anyway (in fact, that's how I do it, and how I achieve Android + CoreELEC + EmuELEC triple boot all on eMMC on s905x4). The u-boot env ``bootcmd`` can be changed simply with the following command through SSH or serial:
```
fw_setenv bootcmd 'if test ${bootfromnand} = 1; then setenv bootfromnand 0; saveenv; else usleep 1000000; run bootfromusb; run bootfromemmc; fi; run storeboot'
```

**Comment by wqmeng** [comment]
Hello,

I have tried your cmd,  if the BOX boot from the USB HDD, the cmd will change the bootcmd on the HDD, so it will not work. 
I also tried to boot from the atv from the S905R4 OS, but it can not run the cmd, sames it do not have that fw_ pkg on it. I do not know if the atv also use uboot as its bootloader. Sorry to say that I never use ubuntu or debian before, Only knows something of CentOS and grub. 

That means if you do not want to re-flash the OS of the TV box, use the TF card is a simple way. If you want to re-flash the OS in the eMMC, you can then change the bootcmd. 

Or you can only put the arm_autoscript.src file on a SD-card, then boot to the bootloader, it will run the script and update the bootcmd? 

And as you said you have already install the 3 OS in the eMMC of your s905x4, can you select the OS when boot such as the Boot menu in a Win PC or Grub2 menu on a linux.  Can the red-line controller of the TV select the OS when the boot menu showing? 

Thank you.

**Comment by wqmeng** [comment]
Also you can try to get the disk size when mk the image, and if the size is over 2TB then use gpt and exFAT else use mbr and FAT32 for the EEROMS partition.
`parted -s -- /dev/$DESTDISK unit B print`

What I know 2TB is "2000409772032", it may different of some HDDs but you can change it to a smaller value for EE to compare.

**Comment by 7Ji** [comment]
@wqmeng 
Please prepare enough background knowledge before digging the problem blindly, I'm not trying to look elitist but you are lacking many essential understanding of the boot process on Amlogic, on ARM, and even on x86-64.

Anything bootable on a post-S9 gen1 Amlogic platform, uses u-boot as bootloader.

If you want to dig about the booting process and manipulate it with ease, try to install Arch Linux on x86-64 EFI with kernel directly as EFISTUB first, without any bootloader, without guided installation. The way it works is very similar to how it works on u-boot (You need to manually set up kernel cmdline for booting, which CE/EE/Armbian/OpenWrt's image just hides it from you with that ``aml_autoscript``)

>sames it do not have that fw_ pkg on it.

The ``fw_printenv`` and ``fw_setenv`` pair are available both under CoreELEC and EmuELEC, I don't understand why you want to do that under Android, Android is locked as fk and don't even try to do anything breaking on it. 

Benifit of doing it under CE/EE is that they uses Amlogic's in-house driver, which has an eMMC driver that'll create a block device ``/dev/env`` so ``fw_`` family does not need manually setting where env is. Where on Ubuntu/Debian (Remember, Armbian is a project name, not a distro name, you create Ubuntu/Debian images with an Armbian build system) / OpenWrt with mainline kernel, there's no such device and you have to manually setting the offset of the ``env`` partition, whose offset can vary a lot depending on the actual EPT partition layout. (Use https://github.com/7Ji/ampart for reading and editting the EPT from eMMC, it is incorporated into EmuELEC as I officialy support it here, but can be built on any machine with a simple ``make``)

> if the size is over 2TB then use gpt and exFAT else use mbr and FAT32 for the EEROMS partition.

This is a naive solution. A disk that has MBR on it can't be eaily converted to GPT loselessly. When ``fs-resize`` is running, you are already booting EmuELEC **on a MBR disk**, we're just one-line close to breaking the disk when doing this: we're already resizing partitions alongside the part-1 (``EMUELEC``) where the kernel and system image resides. Now you want to convert it from MBR to GPT, you'll need to release the part-1 before releasing the whole disk for conversion, and essentially kill our fs-resize target as it is running from there.

And the default FS will be one and only one, it is FAT32 becuase it just works. If you want exFAT you can just create that mark file the same time you prepare DTB. Multiple defaults just confuse the end-users. And the image is never intended to be used on a >2TiB USB drive. The biggest SDXC card on the market is 1TiB and few will try to use a stupidly large yet slow spin drive for the boot drive: You shouldn't do that on x86 (dude, get a SSD for OS, leave HDD for library), why would you do this on the box?

**Comment by wqmeng** [comment]
Hello,

As I know the disk size is limited by the partition table not the file system type, mbr only support 2TB, gpt resolve the problem. Once you use a mbr img on a HDD over 2TB, even you put the ee_fstype as exFat, the EEROMS partition can not use all the disk space. 

We create the mbr or gpt table on the img before we flash it to a real disk. Than means if a real disk is over 2TB, you want to flash the img on it, and the img must use gpt to use full of the space on that HDD, then a selectable option when mkimage will be useful.

As said above, HDD can be over 1TB than SSD, we can store more games, movies on it. The speed is enough for retro games, and can also store many many movies for ATV. 

EE is a very special system focus on retrogames for powerless devices such as TV box.  I can not think that there will be many other usage purpose. If all TV box use u-boot as bootloader, it should be safe to use gpt, though I do not know when and which version of u-boot begin support booting up gpt disk.

Thank you.

**Comment by 7Ji** [comment]
It's a specific requirement for you and few. We won't leave un-maintainable codes in the repo just for the sake of some people might use it. If that's the case, people can just apply your patch before compiling. 

And let's ask a question: How many EE users are out there? Of which how many can compile an EE image? Of which how many will make their changes to their EE source? Of which how many will need a GPT image because they want to use a >=2TiB drive? Of which how many will be lucky enough to get a box whose u-boot actually support GPT? Of which how many, with this extent of knowledge, will still be limited by the partition table in the image, when they can just format a disk and drop files on it freely as documented in https://7ji.github.io/elec/2022/07/08/elec_tarboll_to_image.html?

(And please, you don't seem to understand how I said GPT support is optional, it's optional because the GPT support code is in the source of official u-boot, but manuafacturer can choose whether to include it or not in the build, and many choose not. This has nothing to do with u-boot version, especially Amlogic has its own u-boot fork and many manufacturers then their own. Drop your centric understanding of software versioning because u-boot is open-source and everybody can pull in u-boot and make their changes and call it a day)

> 2TB is "2000409772032"

Mind you, please, I always use TiB in my replies because both MBR and FAT32 are limited to 2TiB = 2\*0x400GiB = 2\*0x100000MiB = 2\*0x40000000KiB = 2\*0x10000000000B = 2199023255552B

EE as a whole will be kept in this way, as we will test and make sure everything works before **every single release**, versioned, test, or private test. We already can't keep up with the growing size of devices being supported. >2TiB big drives are supported, but only, and only suggested for external roms drive. 

And let me remind you if you want to use the drive well: did you ever note you are using ``usb-storage`` driver for your external drive on EE? Plug it to another x86 Linux machine and check its driver, it's probably ``UAS``. The ``UAS`` driver is definitely better suited for big USB drives but guess why we won't enable it by default on CE and EE? Build ``UAS`` and use a ``UAS`` drive for the bootup drive, you'll have some surprise, and further understand why we don't change things that already work
